src/rebar_prv_common_test.erl

```diff
@@ -37,54 +38,352 @@ init(State) ->
-    case Result of
-        {error, Reason} ->
-            {error, {?MODULE, Reason}};
-        ok ->
-            {ok, State}
+    try
+        case setup_ct(State) of
+            {error, {no_tests_specified, Opts}} ->
+                ?WARN("No tests specified in opts: ~p", [Opts]),
+                {ok, State};
+            Opts ->
+                Opts1 = setup_logdir(State, Opts),
+                ?DEBUG("common test opts: ~p", [Opts1]),
+                run_test(State, Opts1)
+        end
+    catch error:Reason -> ?PRV_ERROR(Reason)
     end.
 
 -spec format_error(any()) -> iolist().
+format_error({multiple_dirs_and_suites, Opts}) ->
+    io_lib:format("Multiple dirs declared alongside suite in opts: ~p", [Opts]);
+format_error({bad_dir_or_suite, Opts}) ->
+    io_lib:format("Bad value for dir or suite in opts: ~p", [Opts]);
 format_error({failures_running_tests, {Failed, AutoSkipped}}) ->
     io_lib:format("Failures occured running tests: ~b", [Failed+AutoSkipped]);
 format_error({error_running_tests, Reason}) ->
     io_lib:format("Error running tests: ~p", [Reason]);
-format_error({error_processing_options, Reason}) ->
-    io_lib:format("Error processing options: ~p", [Reason]).
+format_error({error, Reason}) ->
+    io_lib:format("Unknown error: ~p", [Reason]).
+
+%% ===================================================================
+%% Internal functions
+%% ===================================================================
+
+run_test(State, Opts) ->
+    {RawOpts, _} = rebar_state:command_parsed_args(State),
+    Result = case proplists:get_value(verbose, RawOpts, false) of
+        true  -> run_test(Opts);
+        false -> run_test_quiet(Opts)
+    end,
+    ok = rebar_prv_cover:maybe_write_coverdata(State, ?PROVIDER),
+    case Result of
+        ok    -> {ok, State};
+        Error -> Error
+    end.
+
+run_test(Opts) -> handle_results(ct:run_test(Opts)).
 
-run_test(CTOpts, true) ->
-    handle_results(ct:run_test(CTOpts));
-run_test(CTOpts, false) ->
+run_test_quiet(Opts) ->
     Pid = self(),
-    LogDir = proplists:get_value(logdir, CTOpts),
+    LogDir = proplists:get_value(logdir, Opts),
     erlang:spawn_monitor(fun() ->
         {ok, F} = file:open(filename:join([LogDir, "ct.latest.log"]),
                             [write]),
         true = group_leader(F, self()),
-        Pid ! ct:run_test(CTOpts)
+        Pid ! ct:run_test(Opts)
     end),
-    receive Result -> handle_quiet_results(CTOpts, Result) end.
+    receive Result -> handle_quiet_results(Opts, Result) end.
+
+handle_results(Results) when is_list(Results) ->
+    Result = lists:foldl(fun sum_results/2, {0, 0, {0,0}}, Results),
+    handle_results(Result);
+handle_results({_, Failed, {_, AutoSkipped}})
+  when Failed > 0 orelse AutoSkipped > 0 ->
+    ?PRV_ERROR({failures_running_tests, {Failed, AutoSkipped}});
+handle_results({error, Reason}) ->
+    ?PRV_ERROR({error_running_tests, Reason});
+handle_results(_) ->
+    ok.
```

Comment:
I'd probably break this up in two functions -- `sum_results` --> `handle_results`. One is in charge of bundling them together, and the other to react according to the result set.